### DISCUSSION TITLE
trial index for each participant

### DISCUSSION THREAD

**Discussion by EceKayaMPI** [jspsych/jsPsych#865] [Q&A]
Hi, 
Is there a way to pre-define a trial index list for each participant and somehow automate that based on participant number?
Even though no automation option exists, I wish to know options for predefining trial index list, so I can generate a separate html file for each participant. 

Thanks in advance,
Ece

**Comment by jodeleeuw** [comment]
Hi Ece,

Anytime that you want to coordinate across subjects you'll need some kind of back-end script. There isn't a way in JavaScript to determine how many other people have completed/started the experiment so far. 

If you have a back-end script to generate a sequential participant ID then you can certainly automate condition assignment in jsPsych. For illustration purposes, here's an example that assigns a different timeline to participants with even vs. odd IDs:

```js
var participant_id = getParticipantID(); // this would need to communicate with the server

var odd_timeline = [...]; // timeline for odd-IDed participants
var even_timeline = [...]; // timeline for even-IDed participants

if(participant_id % 2 == 0) {
	var timeline = even_timeline;
} else {
	var timeline = odd_timeline;
}

jsPsych.init({
	timeline: timeline
});
```

**Reply by jodeleeuw** [reply]
See this post for some back-end script options: https://github.com/jspsych/jsPsych/discussions/804#discussioncomment-32693

**Reply by EceKayaMPI** [reply]
Hi, 
Thanks a lot for your reply. I still could not fully grasp how the trial loop works. My aim is to present the audio stimuli with different predefined indexes to each participant, but only solution I could think of is to define a 'stimuli list' variable for each person and turn randomisation off. Can you suggest another workaround for this? 

Our institute uses JATOS, but I don't know how to feed such a file there, such that the same list is assigned to a participant, also in multiple sessions. I would appreciate if someone could help me on this as well. 

Cheers,
Ece

**Comment by becky-gilbert** [comment]
Hi @EceKayaMPI!

I'm also a JATOS user :) What I would do is use JATOS' batch session data for this: 
http://www.jatos.org/Session-Data-Three-Types.html
http://www.jatos.org/jatos.js-Reference.html#functions-to-access-the-batch-session

If you want to pre-define a set of trial index lists, you can do that in your study's batch data (Worker & Batch Manager -> Batch Session Data) by creating a JSON object to store each ID and the associated array, e.g.:

```json
{
  "s1": [1,2,3],
  "s2": [3,2,1]
}
```

Then, in your script, if you have the participant's ID you can match that up with the corresponding ID that's saved in your batch data. E.g. if the `id` variable is 's1', then you can do something like this:

```javascript
// Initialize with a default value, just in case the ID does not exist in the batch data
var trial_index_list = [1,2,3]
if (typeof(jatos.batchSession.get(id)) !== "undefined") {
  // This ID exists in the batch data. Get the trial index array for this participant.
  trial_index_list = jatos.batchSession.get(id);
} 
```
Then you should be able to use the `trial_index_list` however you need to in your experiment. Is that something you need help with as well?

**Reply by EceKayaMPI** [reply]
Hi @becky-gilbert 
Thanks for taking time to help me. I followed the steps you suggested, and:
1) in worker&batch manager / new batch / batch session data -> I inserted the example array { "s1": [1,2,3], "s2": [3,2,1] }
2) in the code, for testing if batch thingy works, I added: 
var batchSession = jatos.batchSession.getAll();

but when I run it on a single or multiple worker, I get the error: 
jatos.js:732 Uncaught ReferenceError: jsonpointer is not defined
    at Object.jatos.batchSession.find (jatos.js:732)
    at Object.jatos.batchSession.getAll (jatos.js:719)
    at start?srid=52:39
jatos.batchSession.find @ jatos.js:732
jatos.batchSession.getAll @ jatos.js:719
(anonymous) @ start?srid=52:39

3) I tried to add a variable to the batch session by: 
var promise = jatos.batchSession.add("/subjects/2", {id: 123, name: "Max"});

I get the error: jatos.js:2509 No open batch channel

About the trial index variable: I still don't know how trials are indexed, since I am trained in Matlab or similar languages where I can see and modify the index value. 
But that is fine if I can upload a new list for each participant (and in this case, 'single batch worker' ?) by hand. I can pre-generate the whole stimuli list beforehand. But for that I need to be able to use batch session data, which seems to fail for now.

What do you think may be the cause of the errors?

**Reply by becky-gilbert** [reply]
Hi @EceKayaMPI, sorry for the slow response! I somehow missed this question.

Were you able to solve this problem? If not, would you mind posting this in the [JATOS forum](https://groups.google.com/forum/#!forum/jatos)? I'd be happy to respond there. And once we figure out the JATOS side of things, we can come back to this forum if you have questions about how to use the information in JATOS batch data to set up different trial orders etc. for each participant.
